File: smali_generator/app/src/main/java/com/smali_generator/patches/MetaAiButton.java
package com.smali_generator.patches;

import android.util.Log;

import com.arthooks.ArtHooks;

import com.smali_generator.Hook;
import com.smali_generator.HookCategory;

import java.lang.reflect.Executable;
import java.lang.reflect.Method;
import java.util.ArrayList;


/**
 * Takes Meta AI off the home screen.
 *
 * WhatsApp has put Meta AI on that screen two different ways, and which one a
 * build and an account get is not this patch's to choose, so it removes both:
 *
 * <ul>
 * <li>the <b>tab</b> in the bottom bar, between Updates and Calls. The bar is
 *     drawn from an ordered list of tab ids the app builds at startup, so a tab
 *     whose id is not in the list does not exist. Entries already come and go
 *     there per account -- Updates drops out under a privacy setting, the first
 *     slot is Meta AI or one of two other tabs depending on gating -- so taking
 *     one out is a shape the screen is built to take;</li>
 * <li>the <b>button</b> above the new-chat FAB, which appears as a round fab
 *     carrying the Meta AI ring or as an extended pill with text beside it. The
 *     screen asks one question before it draws either, and before it acts on a
 *     tap: is the Meta AI button on? That is a setting WhatsApp keeps in its own
 *     preferences, so answering no is a state it already has a name for. The
 *     calls tab shows the same button from the same setting but asks the
 *     question itself, with its own copy of the check, so it is answered
 *     separately.</li>
 * </ul>
 *
 * No half depends on another, and each is installed in its own try/catch: a
 * build that has only some of these entry points still loses them.
 *
 * Hooking the chat list's own "do you want the pill" gate instead was the
 * earlier mistake. It is only one of the four callers of the real gate, so the
 * round fab -- the form most accounts get -- was left in place, and on a device
 * that was never offered the pill the patch logged success and changed nothing.
 */
public class MetaAiButton implements Hook {

    private static final String TAG = "PATCH";

    /**
     * The Meta AI tab's id, resolved at load.
     *
     * Left at NO_TAB when the artifact did not resolve, which is also the only
     * path on which the list is not hooked at all -- so the value is never used
     * to remove a tab that was not identified.
     */
    private static final int NO_TAB = -1;
    private static volatile int metaAiTab = NO_TAB;

    /** Volatile so "once" holds however the home screen is rebuilt. */
    private static volatile boolean loggedTabs;
    private static volatile boolean loggedButton;
    private static volatile boolean loggedCalls;

    /** native on purpose: a body here would be inlined into the hook and the backup would
     * silently answer for the original. ArtHooks rewrites its entry point. */
    static native ArrayList tabs_backup(Object thiz);

    static ArrayList tabs_hook(Object thiz) {
        ArrayList tabs = tabs_backup(thiz);
        try {
            // The app's own list, minus one entry: everything downstream --
            // which tab is selected, what the bar draws, what the pager holds --
            // is derived from it, so there is nothing else to keep in step.
            boolean removed = tabs != null && tabs.remove(Integer.valueOf(metaAiTab));
            if (!loggedTabs) {
                loggedTabs = true;
                Log.i(TAG, "MetaAiButton: home tabs " + tabs + (removed
                        ? ", Meta AI removed" : ", no Meta AI tab to remove"));
            }
        } catch (Throwable t) {
            // The home screen is being built; a throw here would take it down.
            if (!loggedTabs) {
                loggedTabs = true;
                Log.e(TAG, "MetaAiButton: could not edit the tab list: " + t);
            }
        }
        return tabs;
    }

    /** native on purpose: a body here would be inlined into the hook and the backup would
     * silently answer for the original. ArtHooks rewrites its entry point. */
    static native boolean button_gate_backup(Object thiz);

    static boolean button_gate_hook(Object thiz) {
        if (!loggedButton) {
            loggedButton = true;
            Log.i(TAG, "MetaAiButton: the Meta AI button is off on the chats screen"
                    + "; WhatsApp would have shown it: " + wanted(thiz, true));
        }
        return false;
    }

    /** native on purpose: a body here would be inlined into the hook and the backup would
     * silently answer for the original. ArtHooks rewrites its entry point. */
    static native boolean calls_gate_backup(Object thiz);

    static boolean calls_gate_hook(Object thiz) {
        if (!loggedCalls) {
            loggedCalls = true;
            Log.i(TAG, "MetaAiButton: the Meta AI button is off on the calls screen"
                    + "; WhatsApp would have shown it: " + wanted(thiz, false));
        }
        return false;
    }

    /**
     * What WhatsApp's own gate would have said, for the log line only.
     *
     * It is the difference between "the button was taken away" and "this
     * account was never offered one", which is otherwise indistinguishable from
     * outside -- and telling those apart is the whole reason this hook was wrong
     * the first time. Safe to ask: both originals only read a setting and a
     * gating flag. Guarded anyway, since a throw here would land in a call
     * WhatsApp makes while building the screen.
     */
    private static String wanted(Object thiz, boolean chats) {
        try {
            return String.valueOf(chats ? button_gate_backup(thiz) : calls_gate_backup(thiz));
        } catch (Throwable t) {
            return "unknown (" + t + ")";
        }
    }

    public String id() {
        return "meta_ai_button";
    }

    public String title() {
        return "Remove Meta AI from the home screen";
    }

    public String description() {
        return "Takes the Meta AI tab out of the bottom bar, and the Meta AI button off the chats and calls screens.";
    }

    public HookCategory category() {
        return HookCategory.INTERFACE;
    }

    public void load() {
        remove_tab();
        remove_button();
        remove_calls_button();
    }

    private void remove_tab() {
        try {
            // Parsed rather than inlined as a literal: an unsubstituted
            // placeholder then fails here, before anything is hooked, instead of
            // becoming a tab id nothing matches.
            metaAiTab = Integer.parseInt("{{META_AI_TAB_ID}}");
            Class<?> tabs = Class.forName("{{META_AI_TABS_CLASS_NAME}}");

            Executable builder = ArtHooks.find_function(tabs,
                    "{{META_AI_TABS_METHOD_NAME}}", "{{META_AI_TABS_METHOD_SIG}}");
            Method replacement = MetaAiButton.class.getDeclaredMethod("tabs_hook", Object.class);
            Method original = MetaAiButton.class.getDeclaredMethod("tabs_backup", Object.class);
            boolean hooked = ArtHooks.hook_function(builder, replacement, original);

            Log.i(TAG, "MetaAiButton: tab " + metaAiTab + " hooked on " + tabs.getName()
                    + ".{{META_AI_TABS_METHOD_NAME}}, hooked=" + hooked);
        } catch (Throwable t) {
            Log.e(TAG, "MetaAiButton: the Meta AI tab was left in place: " + t);
        }
    }

    private void remove_button() {
        try {
            Class<?> owner = Class.forName("{{META_AI_BUTTON_GATE_CLASS_NAME}}");

            Executable gate = ArtHooks.find_function(owner,
                    "{{META_AI_BUTTON_GATE_METHOD_NAME}}", "{{META_AI_BUTTON_GATE_METHOD_SIG}}");
            Method replacement = MetaAiButton.class.getDeclaredMethod("button_gate_hook", Object.class);
            Method original = MetaAiButton.class.getDeclaredMethod("button_gate_backup", Object.class);
            boolean hooked = ArtHooks.hook_function(gate, replacement, original);

            Log.i(TAG, "MetaAiButton: button gate hooked on " + owner.getName()
                    + ".{{META_AI_BUTTON_GATE_METHOD_NAME}}, hooked=" + hooked);
        } catch (Throwable t) {
            Log.e(TAG, "MetaAiButton: the Meta AI button was left on the chats screen: " + t);
        }
    }

    private void remove_calls_button() {
        try {
            Class<?> calls = Class.forName("{{META_AI_CALLS_GATE_CLASS_NAME}}");

            Executable gate = ArtHooks.find_function(calls,
                    "{{META_AI_CALLS_GATE_METHOD_NAME}}", "{{META_AI_CALLS_GATE_METHOD_SIG}}");
            Method replacement = MetaAiButton.class.getDeclaredMethod("calls_gate_hook", Object.class);
            Method original = MetaAiButton.class.getDeclaredMethod("calls_gate_backup", Object.class);
            boolean hooked = ArtHooks.hook_function(gate, replacement, original);

            Log.i(TAG, "MetaAiButton: calls gate hooked on " + calls.getName()
                    + ".{{META_AI_CALLS_GATE_METHOD_NAME}}, hooked=" + hooked);
        } catch (Throwable t) {
            Log.e(TAG, "MetaAiButton: the Meta AI button was left on the calls screen: " + t);
        }
    }

    public void unload() {
        Log.i(TAG, "MetaAiButton: Patch unloaded");
    }
}

File: artifactory_generator/meta_ai_button.py
import re
from stitch.artifactory_generator.SimpleArtifactoryFinder import SimpleArtifactoryFinder

from artifactory_generator.smali import CLASS_RE

# The setting WhatsApp's own Settings > Chats screen writes for its Meta AI
# button. It is a SharedPreferences key, so renaming it would strand the stored
# value of every user who has ever touched that switch: it does not move.
BUTTON_PREF = 'bonsai_meta_ai_button_setting_enabled'
PREF_RE = re.compile(r'const-string(?:/jumbo)? [vp]\d+, "%s"' % BUTTON_PREF)
GET_BOOLEAN_RE = re.compile(r'->getBoolean\(Ljava/lang/String;Z\)Z')

METHOD_RE = re.compile(
    r'^\.method (?P<modifiers>[\w ]*?)(?P<name>[\w$]+)(?P<sig>\([^)\n]*\)[\w/$;\[]+)\n'
    r'(?P<body>.*?)^\.end method$',
    re.MULTILINE | re.DOTALL)
GATE_SIG = '()Z'


def _reads_the_pref(method) -> bool:
    body = method.group('body')
    return (method.group('sig') == GATE_SIG
            and PREF_RE.search(body) is not None
            and GET_BOOLEAN_RE.search(body) is not None)


class MetaAiButtonFinder(SimpleArtifactoryFinder):
    """Finds the gate for the Meta AI button on the chats screen.

    The button has two forms -- a round fab carrying the Meta AI ring, and an
    extended pill with text beside it -- and the home screen's fab controller
    picks between them. Which one an account sees is decided elsewhere, but
    *whether* it sees either is decided here: this one predicate is consulted by
    the chat list before it hands out the ring drawable, before it hands out the
    ring resource id, before it answers whether it wants the pill, and before it
    acts on a tap. Returning false is the state of every account whose Meta AI
    button is off, so it needs no new handling anywhere.

    Anchoring on the chat list's own methods instead was the earlier mistake and
    is what this replaces. The pill's gate is only one of the four callers, so
    hooking it left the round fab -- the form most accounts get -- untouched,
    and on a device where the pill was never offered the patch looked installed
    and did nothing.

    The class is found by what it is for: it declares exactly one method, a
    ``()Z`` that reads {BUTTON_PREF} out of SharedPreferences. Nothing else in
    the app has that shape, while the class itself was renamed in every build
    examined -- ``X.16t``, ``X.17f``, ``X.118``, ``X.12O``, ``X.13z``, ``X.1LK``.
    That last one is why the name is read with the shared ``CLASS_RE``: stitch's
    own regex made it ``K``.

    Deliberately not hooked: WhatsApp's own Settings > Chats screen, which reads
    the same key directly to draw its switch. It keeps showing the user's real
    setting rather than one this patch invented.

    Verified to resolve the gate, and nothing else, on 2.26.17.72, 2.26.27.85,
    2.26.29.74, 2.26.33.76, 2.26.36.71 and 2.26.37.74.
    """

    def __init__(self, args):
        super().__init__(args)
        self.is_once = True
        self.is_found = False

    def class_filter(self, class_data: str) -> bool:
        return PREF_RE.search(class_data) is not None

    def extract_artifacts(self, artifacts: dict, class_data: str) -> None:
        methods = list(METHOD_RE.finditer(class_data))
        # One method and nothing else: the class exists to answer this. The
        # chat list and the settings screen read the same key inside much larger
        # classes, and neither is what should be hooked.
        if len(methods) != 1 or not _reads_the_pref(methods[0]):
            return
        class_match = CLASS_RE.search(class_data)
        if class_match is None:
            return
        artifacts['META_AI_BUTTON_GATE_CLASS_NAME'] = class_match.groupdict().get('name').replace('/', '.')
        artifacts['META_AI_BUTTON_GATE_METHOD_NAME'] = methods[0].group('name')
        artifacts['META_AI_BUTTON_GATE_METHOD_SIG'] = GATE_SIG
        self.is_found = True


class MetaAiCallsButtonFinder(SimpleArtifactoryFinder):
    """Finds the gate for the Meta AI button on the calls screen.

    The calls tab shows the same button from the same setting, but asks the
    question itself rather than through the shared gate
    {MetaAiButtonFinder} pins: it reads the two fields off that class
    inline and pairs them with a different availability check. So it survives a
    hook on the shared gate, and without this the button does not go away -- it
    moves one tab over.

    Anchored on the same SharedPreferences key, inside a class whose name is in
    an unobfuscated package. Within it the method is the only ``()Z`` that reads
    the key, which is how its name is left an output: it was ``A0V``, ``A0X``,
    ``A0Y`` and ``A0Z`` across four builds.

    Verified on 2.26.27.85, 2.26.29.74, 2.26.33.76 and 2.26.36.71. 2.26.17.72
    has no Meta AI button on the calls tab and nothing here to find, so the
    finder correctly does not fire on it.
    """

    FRAGMENT_CLASS = 'com/whatsapp/calling/ui/callhistory/view/CallsHistoryFragment'

    def __init__(self, args):
        super().__init__(args)
        self.is_once = True
        self.is_found = False

    def class_filter(self, class_data: str) -> bool:
        if PREF_RE.search(class_data) is None:
            return False
        match = CLASS_RE.search(class_data)
        return match is not None and match.groupdict().get('name') == self.FRAGMENT_CLASS

    def extract_artifacts(self, artifacts: dict, class_data: str) -> None:
        gates = [method for method in METHOD_RE.finditer(class_data) if _reads_the_pref(method)]
        if len(gates) != 1:
            return
        artifacts['META_AI_CALLS_GATE_CLASS_NAME'] = self.FRAGMENT_CLASS.replace('/', '.')
        artifacts['META_AI_CALLS_GATE_METHOD_NAME'] = gates[0].group('name')
        artifacts['META_AI_CALLS_GATE_METHOD_SIG'] = GATE_SIG
        self.is_found = True
